### PR TITLE
feat(training)!: make LR scheduler configurable through Hydra and clean up optimization config

### LINKED ISSUE #653: Configurable LR Schedulers
### Is your feature request related to a problem? Please describe.

Currently the learning rate scheduler is hard-coded. Similar to https://github.com/ecmwf/anemoi-core/pull/588 we should have arbitrary pytorch learning rate schedulers available to users. 

### Describe the solution you'd like

Instantiate schedulers directly, for instance as in [here](https://github.com/ashleve/lightning-hydra-template/blob/bddbc24b82ab6ccfa6243e815a49dc5bfe8d4144/configs/model/mnist.yaml#L9-L14): 

```
scheduler:
  _target_: torch.optim.lr_scheduler.ReduceLROnPlateau
  _partial_: true
  mode: min
  factor: 0.1
  patience: 10
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by matschreiner**
# Summary

The LR scheduler, optimizer, and learning rate were previously scattered across every training config preset as duplicated inline blocks, with the scheduler hardcoded to `CosineLRScheduler` in `configure_optimizers`. This PR consolidates all optimization configuration into a dedicated `optimization` config group, makes both the optimizer and scheduler fully configurable via the standard Hydra `_target_` pattern, and cleans up the code paths that read these settings.

# Changes

**`training/config/training/optimization/default.yaml`** (new)
- Single source of truth for `lr`, `optimizer`, `lr_scheduler`, and `pl_lr_scheduler`
- `lr_scheduler` uses `_target_: timm.scheduler.CosineLRScheduler` — preserves existing default behaviour
- `pl_lr_scheduler` controls Lightning wrapper settings (interval, monitor, etc.) independently of the scheduler itself; defaults to `interval: step`

**`training/config/training/optimization/optimizer/`** (new)
- `adamw.yaml`, `ademamix.yaml`, `zero.yaml` — ready-made optimizer presets to drop in via override

**`training/config/training/{default,autoencoder,ensemble,lam,multi,stretched}.yaml`**
- Adds `- optimization: default` to defaults list
- Removes duplicated inline `optimizer:` and `lr:` blocks from all preset configs

**`training/config/training/diffusion.yaml`**
- Adds `- optimization: default`; overrides only `optimization.optimizer` inline (custom `weight_decay`/`eps`); inherits `lr`, `lr_scheduler`, and `pl_lr_scheduler` from the default

**`training/train/tasks/base.py`**
- `configure_optimizers` now instantiates both optimizer and scheduler via `hydra.utils.instantiate`; if `lr_scheduler` is absent or null, returns the optimizer alone (constant LR)
- Removed private `_create_optimizer_from_config` and `_create_scheduler` methods
- `lr_scheduler_step` generalised to handle both timm-style schedulers (`step_update`) and standard PyTorch schedulers
- `self.lr` renamed to `self.effective_lr`; reads from `config.training.optimization.lr`

**`training/train/tasks/ensforecaster.py`**
- Same `self.lr` → `self.effective_lr` / `config.training.lr.rate` → `config.training.optimization.lr` update

**`training/train/train.py`**
- `config.training.lr.rate` / `config.training.lr.iterations` → `config.training.optimization.lr` / `config.training.max_steps`

**`training/schemas/training.py`**
- Removes `LR` class; adds `OptimizationSchema` grouping `lr`, `optimizer`, `lr_scheduler`, `pl_lr_scheduler`; `BaseTrainingSchema.optimization` replaces the old `lr` + `optimizer` fields

**`training/tests/unit/train/test_optimizer.py`**
- Tests rewritten against the `configure_optimizers` public API; private helpers no longer tested directly

**`training/tests/integration/config/testing_modifications.yaml`**
- `training.lr.{warmup,rate}` → `training.optimization.lr`

# Usage

Disable scheduling (constant LR):
```yaml
training:
  optimization:
    lr_scheduler: null
```

Switch to a different scheduler:
```yaml
training:
  optimization:
    lr_scheduler:
      _target_: torch.optim.lr_scheduler.ReduceLROnPlateau
      mode: min
      factor: 0.5
      patience: 5
    pl_lr_scheduler:
      monitor: val_loss
      interval: epoch
```

Switch to AdEMAMix optimizer:
```yaml
training:
  optimization:
    optimizer:
      _target_: anemoi.training.optimizers.AdEMAMix.AdEMAMix
      betas: [0.9, 0.95, 0.9999]
      alpha: 8.0
```
or use the preset file: `defaults: [optimization/optimizer: ademamix]`

### CODE REVIEW

File: training/tests/unit/schemas/test_allowed_extra_keys.py
Comment (VeraChristina) [inline training/tests/unit/schemas/test_allowed_extra_keys.py:1]: Why remove this test?

File: training/src/anemoi/training/schemas/training.py [hunk 1]
```diff
@@ -64,30 +62,17 @@ class Rollout(BaseModel):
     "Maximum number of rollouts."
 
 
-class LR(BaseModel):
-    """Learning rate configuration.
+class OptimizationSchema(BaseModel):
+    """Optimizer and LR scheduler configuration."""
```
Comment (VeraChristina) [inline training/src/anemoi/training/schemas/training.py:94]: How do we handle validation of the optimizer config, and how do we allow additional keys for the optimiser (related to the test that was removed)?

**Review by VeraChristina — COMMENTED** [review]
Thanks for adding this! Closes https://github.com/ecmwf/anemoi-core/issues/653

Two questions related to the optimizer schema in the comments. For context, this has led to problems earlier. The test was added to avoid regressions.

Have you run the integration tests locally? The last run I found was failing in an internal check of the LR scheduler but it might be outdated: https://github.com/ecmwf/anemoi-core/actions/runs/23288208725/job/67716841148

File: training/src/anemoi/training/train/tasks/base.py [hunk 2]
```diff
@@ -1058,62 +1056,57 @@ def validation_step(self, batch: dict[str, torch.Tensor], batch_idx: int) -> Non
...
+        else:
+            scheduler.step()
 
     def on_train_epoch_end(self) -> None:
         pass
 
-    def configure_optimizers(self) -> tuple[list[torch.optim.Optimizer], list[dict[str, Any]]]:
+    def configure_optimizers(
+        self,
+    ) -> OptimizerLRScheduler:
```
Comment (MeraX) [inline training/src/anemoi/training/train/tasks/base.py:1084]: Could you check this return type?

Comment (matschreiner) [reply training/src/anemoi/training/train/tasks/base.py:1084]: Do you mean whether it's the correct type? It is PLs own return type for the configure_optimizers interface. Or are you talking about something else?

File: training/tests/integration/config/testing_modifications.yaml [hunk 3]
```diff
@@ -44,6 +44,5 @@ model:
 
 training:
   max_epochs: 2
-  lr:
-    warmup: 1
```
Comment (VeraChristina) [inline training/tests/integration/config/testing_modifications.yaml:48]: why remove this modification?

File: training/tests/unit/schemas/test_allowed_extra_keys.py [hunk 4]
```diff
@@ -1,13 +1,3 @@
-# (C) Copyright 2026- Anemoi contributors.
-#
-# This software is licensed under the terms of the Apache Licence Version 2.0
-# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
-#
-# In applying this licence, ECMWF does not waive the privileges and immunities
-# granted to it by virtue of its status as an intergovernmental organisation
-# nor does it submit to any jurisdiction.
-
-
```
Comment (VeraChristina) [inline training/tests/unit/schemas/test_allowed_extra_keys.py:10]: why remove the license?

File: training/src/anemoi/training/config/training/stretched.yaml [hunk 5]
```diff
@@ -160,15 +158,6 @@ rollout:
 max_epochs: null
 max_steps: 150000
 
-lr:
-  warmup: 1000 # number of warmup iterations
-  rate: 0.625e-4 #local_lr
```
Comment (VeraChristina) [inline training/src/anemoi/training/config/training/stretched.yaml:165]: I can't seem to find this default setting in the new layout

File: training/src/anemoi/training/config/training/diffusion.yaml [hunk 6]
```diff
@@ -148,16 +149,6 @@ rollout:
 max_epochs: null
 max_steps: 1000000
 
-lr:
-  warmup: 1000 # number of warmup iterations
-  rate: 0.625e-4 #local_lr
```
Comment (VeraChristina) [inline training/src/anemoi/training/config/training/diffusion.yaml:153]: I can't seem to find this default setting in the new layout

**Review by VeraChristina — COMMENTED** [review]
Thanks for fixing the tests! Generally looking good to merge but I can't seem to find the changes from Simon's PR https://github.com/ecmwf/anemoi-core/pull/995

File: training/src/anemoi/training/config/training/diffusion.yaml (hunk 6 above)
Comment (matschreiner) [reply training/src/anemoi/training/config/training/diffusion.yaml:153]: It was renamed to `t_initial` to match the signature of timm's CosineLrScheduler

File: training/tests/unit/schemas/test_allowed_extra_keys.py (hunk 4 above)
Comment (matschreiner) [reply training/tests/unit/schemas/test_allowed_extra_keys.py:10]: Just saw it wasn't there in test_expand_paths.py which was it's neighbor.

File: training/src/anemoi/training/config/training/diffusion.yaml (hunk 6 above)
Comment (VeraChristina) [reply training/src/anemoi/training/config/training/diffusion.yaml:153]: ah, sorry, this was ambiguous. Yes, `steps` was renamed to `t_initial`. How about `rate`?

File: training/tests/integration/config/testing_modifications.yaml (hunk 3 above)
Comment (matschreiner) [reply training/tests/integration/config/testing_modifications.yaml:48]: Right, this is missing! Thanks, added it now :)

File: training/src/anemoi/training/config/training/diffusion.yaml (hunk 6 above)
Comment (matschreiner) [reply training/src/anemoi/training/config/training/diffusion.yaml:153]: I moved it to the top-level config. Surprisingly it isn't a lr scheduler concern. Technically it belongs to the optimizer, but since we are scaling it based on hardware and we are really passing the effective_lr to the optimizer I wanted it on the top level of the config.

**Review by VeraChristina — APPROVED** [review]
Thanks for addressing the feedback, LGTM!
